tools/generate_linux_syscalls.zig

```diff
@@ -48,710 +48,631 @@ fn isReservedNameOld(name: []const u8) bool {
-            const fixed_name = if (stdlib_renames.get(name)) |fixed| fixed else name;
-
-            try writer.print("    {p} = {s},\n", .{ zig.fmtId(fixed_name), number });
+const default_args: []const []const u8 = &.{
+    "-E",
+    // -dM is cleaner, but -dD preserves iteration order.
+    "-dD",
+    // No need for line-markers.
+    "-P",
+    "-nostdinc",
+    // Using -I=[dir] includes the zig linux headers, which we don't want.
+    "-Itools/include",
+    "-Itools/include/uapi",
+    // Output the syscall in a format we can easily recognize.
+    "-D __SYSCALL(nr, nm)=zigsyscall nm nr",
+};
+
+const ProcessPreprocessedFileFn = *const fn (bytes: []const u8, writer: anytype) anyerror!void;
+const ProcessTableBasedArchFileFn = *const fn (
+    bytes: []const u8,
+    filters: Filters,
+    writer: anytype,
+    optional_writer: anytype,
+) anyerror!void;
+
+const FlowControl = enum {
+    @"break",
+    @"continue",
+    none,
+};
+
+const Filters = struct {
+    abiCheck: ?*const fn (abi: []const u8) FlowControl,
+    fixedName: ?*const fn (name: []const u8) []const u8,
+    isReservedNameOld: ?*const fn (name: []const u8) bool,
+};
+
+const AbiCheckFn = *const fn (abi: []const u8) FlowControl;
+
+fn getAbiCheckFunction(comptime b: []const u8, comptime flow: FlowControl) AbiCheckFn {
+    return struct {
+        pub fn abiCheck(abi: []const u8) FlowControl {
+            if (mem.eql(u8, abi, b)) return flow;
+            return .none;
         }
+    }.abiCheck;
+}
 
-        try writer.writeAll("};\n\n");
-    }
-    {
-        try writer.writeAll(
-            \\pub const Arm = enum(usize) {
-            \\    const arm_base = 0x0f0000;
-            \\
-            \\
-        );
-
-        const table = try linux_dir.readFile("arch/arm/tools/syscall.tbl", buf);
-        var lines = mem.tokenizeScalar(u8, table, '\n');
-        while (lines.next()) |line| {
-            if (line[0] == '#') continue;
-
-            var fields = mem.tokenizeAny(u8, line, " \t");
-            const number = fields.next() orelse return error.Incomplete;
-            const abi = fields.next() orelse return error.Incomplete;
-            if (mem.eql(u8, abi, "oabi")) continue;
-            const name = fields.next() orelse return error.Incomplete;
-            const fixed_name = if (stdlib_renames.get(name)) |fixed| fixed else name;
-
-            try writer.print("    {p} = {s},\n", .{ zig.fmtId(fixed_name), number });
-        }
+fn fixedName(name: []const u8) []const u8 {
+    return if (stdlib_renames.get(name)) |fixed| fixed else name;
+}
 
-        // TODO: maybe extract these from arch/arm/include/uapi/asm/unistd.h
-        try writer.writeAll(
+const ArchInfo = union(enum) {
+    tableBasedArch: struct {
+        name: []const u8,
+        enum_name: []const u8,
+        file_path: []const u8,
+        header: ?[]const u8,
+        extra_values: ?[]const u8,
+        processFile: ProcessTableBasedArchFileFn,
```

Comment:
```suggestion
        process_file: ProcessTableBasedArchFileFn,
```

Likewise in the other case below.